### PR TITLE
Improve Tree Generator UI

### PR DESCRIPTION
**Pull request by brather1ng**
These are mainly changes that should make the Tree Generator easier to understand and find, no functionality changes.

- Move "Skill Tagged Nodes" from menu to the generator settings window.
The new "Tagged Nodes" tab contains some explanation as it doesn't
need settings. Settings that don't affect it are disabled when that tab
is selected.
- Change the generator settings window to be fixed width. No more
enlarging when long attribute names are added as constraints.
- Add explanation tooltips to many of the settings.
- Move "Nodes" sub menu to a new top level menu.

Doing this as a Pull Request to give you the chance to comment on the usefulness (mainly @MLanghof as you actually use this a lot).

Also, the "Tagged Nodes" tab still has space for more explanation text. I'd appreciate suggestions on what might be useful to know as a user.

### CODE REVIEW

**Comment by EmmittJ** [comment]
I guess I don't mind the change, but this seems to add another window for no real reason? I am kind of confused why it has been taken from 2 clicks to 3 + a new window.

**Comment by brather1ng** [comment]
The window and everything was already there. I just moved "Skill Tagged Nodes" to it. That was the plan since this was initially added (I had both at first). (also, it's one click less now, as "Nodes" is now a top level menu entry)

I could just add the "Skill Tagged Nodes" menu entry back. But it should also be available from the generator window to make the connection between them clearer and have some space for explanation (see e.g. #369).

**Comment by EmmittJ** [comment]
Ah, ok.  I think it is fine, and the only reason to add it back is if it takes you to the new tab

**Comment by MLanghof** [comment]
I recently set up the new VS version on my PC but had to ditch it (the PC) since I moved to a different country. I'll see to getting the 2015 version on this tablet (multi-GB downloads via wireless, yay), so bear with me for a while. I'll make sure to check it out and let you know, but I suppose you can merge this anyway - if there's anything that should be amended then that shouldn't be a problem.

**Comment by brather1ng** [comment]
Thanks, I'll upload a compiled version tomorrow so you don't have to download VS.

**Comment by MLanghof** [comment]
Well, installing it actually went faster (and easier) than I would've expected. I think I can already take a look right now.

**Comment by MLanghof** [comment]
Here's some things that caught my eye:

- "Reset Settings" puts you back in the "Tagged Nodes" tab, I doubt that's what one would intend to happen.

- I would reword "do not expect optimal solutions from the 'Advanced' generator" to something like "the resulting solutions may deviate slightly from the optimum". I know that "not optimal" technically means exactly that, but it leaves a different impression to the average reader. And in any case, the input constraints are only an abstraction of what the user wants, so the nature of our weighting function(s) likely already introduces more significant deviations from what would be the optimal solution the user is looking for.

- I've never had a use for the checkboxes on the right - "Only search in current tree" is an arbitrary restriction that I can't imagine a workflow for (well, if there was a way to skill all nodes in an area of the tree then maybe that would work, but I don't see anybody clicking 200 nodes manually to limit the generator to that area), and I'm not under the impression the algorithm really needs an initial tree in virtually any case (it's very flexible in the early generations anyway). In either case I'd be more worried about shoehorning the GA into a local optimum than preventing a complete failure (which we've never even observed afaik), and the second checkbox certainly is redundant for the "Tagged Nodes" solver (it really doesn't need hints).

Given the above, I'd imagine throwing the level etc. edits and the "Reset Settings" button (and the two checkboxes on the right if you want to keep them) into the Advanced tab would make the UI more intuitive (also eliminating the need for the corresponding paragraph in the mini-help).
So in the end the only controls outside the tabs would be the include/exclude checkboxes and the run button. The "Tagged Nodes" tab would have nothing but the explanatory text, and everything else goes in the "Advanced Tab" (because it only has an effect there\*).

\*Well, if we keep "only search in current tree" then I guess it would/should affect both generators. Would like to hear your opinion.

**Comment by brather1ng** [comment]
"Only search in current tree" was meant as a possibility to create leveling trees, but I have no idea if anyone would use it for that. It's probably simpler to create leveling trees manually. I'll just remove both of them for now, makes both UI and code simpler.

I also agree with your other points. Thanks for checking it out!

**Comment by brather1ng** [comment]
All done and also extended the explanation text.

**Comment by MLanghof** [comment]
Alright, looks good. I feel like the settings tab looks slightly funky (the space between the group boxes), but that's not major.

I also noticed that "include Check-tagged nodes" has to always be enabled for the "Tagged Nodes" tab, otherwise it doesn't make any sense. So that one should probably also be ferried over to the Advanced tab. The last remaining checkbox does have its uses, but it will feel slightly out of place. I don't really know what we should do with it...

**Comment by brather1ng** [comment]
Yeah, well it somewhat boils done to having the same as before plus help text. I do like this more because the relationship between the two generators is obvious this way. As opposed to people unnecessarily running the advanced one as before.

I would like to keep the two checkboxes avaiable for both, mostly so "Tagged Nodes" doesn't look completely deserted.

Regarding it getting in the way, I did like this more than the three clicks away menu entry when I had to use it a ton for testing. Also, you can somewhat put it out of the way by moving it to the bottom of the window until only the run button is reachable.

What I could do is adding keyboard shortcuts for both generators so they can be used without the window being open.

> I feel like the settings tab looks slightly funky (the space between the group boxes), but that's not major.

Yeah, I experimented with that but didn't came up with a good solution.

**Comment by MLanghof** [comment]
> Also, you can somewhat put it out of the way by moving it to the bottom of the window until only the run button is reachable

That was also the workaround I came up with, but surely that's not what UI design should aim for? :P
But yeah, a shortcut would solve that. Then again, if it still requires the Skill Tree Generator window to be open, it's not that much of a gain (because you'd still have to cram the window somewhere every time you work with it).

I'm not sold on keeping "include Check-tagged nodes" available for both generators. I guess it could be disabled while on "Tagged Nodes", but then it'd have to be "frozen" in checked state (undoing the choice the user had for the advanced tab) or be ignored (which would be confusing) or frozen in the wrong state (which would be very bad).

I can get behind the value of juxtaposing both generators for the reasons you mentioned. It just brings some workflow oddities with it.

**Comment by brather1ng** [comment]
This should've been the last commit now.

Both generators have menu entries (that make their relation to the "Skill Tree Generator" window clear) and keyboard shortcuts now. Those don't open the generator window but use the settings from it.

There are no general settings anymore. "Tagged Nodes" has just the Cross-tagged nodes checkbox. Run buttons are also in the tabs.

New users should see the explanation text as the direct menu entries aren't clear without context on purpose. Others don't need to ever see it again as the opened tab is saved and "Tagged Nodes" will probably be used only through the direct menu entry by them.

**Comment by MLanghof** [comment]
Wow, this is awesome! I didn't even think about making the functionalities accessible through both ways, but it elegantly solves both the "wtf does this do?" side and the "don't clutter my UI!" one. Brilliant! :D

**Comment by brather1ng** [comment]
Thanks for all your input, wouldn't have gotten there without it!
